src/antennachecker/include/antennachecker/AntennaChecker.hh

```diff
@@ -0,0 +1,178 @@
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include <tcl.h>
+#include <map>
+#include "opendb/db.h"
+#include "opendb/dbWireGraph.h"
+
+struct PARinfo
+{
+    std::pair<odb::dbWireGraph::Node*, std::vector<odb::dbWireGraph::Node*>> WirerootNode;
+    long unsigned int num_iterms;
+    double wire_area;
+    double side_wire_area;
+    double iterm_areas[2];
+    double PAR_value;
+    double PSR_value;
+    double diff_PAR_value;
+    double diff_PSR_value;
+};
+
+struct ARinfo
+{
+    odb::dbWireGraph::Node * WirerootNode;
+    odb::dbWireGraph::Node * GateNode;
+    bool violated_net;
+    double PAR_value;
+    double PSR_value;
+    double diff_PAR_value;
+    double diff_PSR_value;
+    double CAR_value;
+    double CSR_value;
+    double diff_CAR_value;
+    double diff_CSR_value;
+    double diff_area;
+};
+
+struct ANTENNAmodel
+{
+    odb::dbTechLayer* layer;
+
+    double metal_factor;
+    double diff_metal_factor;
+
+    double cut_factor;
+    double diff_cut_factor;
+
+    double side_metal_factor;
+    double diff_side_metal_factor;
+
+    double minus_diff_factor;
+    double plus_diff_factor;
+    double diff_metal_reduce_factor;
+
+    ANTENNAmodel& operator =(const ANTENNAmodel& am)
+    {
+        metal_factor = am.metal_factor;
+        diff_metal_factor = am.diff_metal_factor;
+        cut_factor = am.cut_factor;
+        diff_cut_factor = am.diff_cut_factor;
+        side_metal_factor = am.side_metal_factor;
+        diff_side_metal_factor = am.diff_side_metal_factor;
+        minus_diff_factor = am.minus_diff_factor;
+        plus_diff_factor = am.plus_diff_factor;
+        diff_metal_reduce_factor = am.diff_metal_reduce_factor;
+
+        return *this;
+    }
+
+
+};
+
+namespace antennachecker {
+
+using odb::dbNet;
+using odb::dbWire;
+using odb::dbSWire;
+using odb::dbInst;
+using odb::dbWireGraph;
+using odb::dbTechLayer;
+using odb::dbITerm;
+
+class AntennaChecker
+{
+public:
+    AntennaChecker();
+    ~AntennaChecker();
+    void init(Tcl_Interp *tcl_interp,
+	    odb::dbDatabase *db);
+
+    void set_verbose( bool verbose );
+    void set_net_name( std::string net_name );
+    void set_route_level( int route_level );
+    
+    double defdist( int value );
+    double defdist( uint value );
+
+    std::pair<dbWireGraph::Node*, std::vector<dbWireGraph::Node*>> find_segment_root(std::pair<dbWireGraph::Node*, std::vector<dbWireGraph::Node*>> node_info, int wire_level );
```

Comment:
some typedefs would be appropriate to prevent copying complicated types like this

Reply:
Done. See new commit.